src/BlockTypes/SingleProduct.php

```diff
@@ -0,0 +1,139 @@
+		add_filter( 'render_block_context', array( $this, 'update_context' ), 10, 3 );
+	}
+
+
+	/**
+	 * Render the Single Product block
+	 *
+	 * @param array    $attributes Block attributes.
+	 * @param string   $content Block content.
+	 * @param WP_Block $block Block instance.
+	 *
+	 * @return string Rendered block output.
+	 */
+	protected function render( $attributes, $content, $block ) {
+		$classname = $attributes['className'] ?? '';
+
+		$html = sprintf(
+			'<div class="wp-block-woocommerce-single-product %1$s">
+				%2$s
+			</div>',
+			esc_attr( $classname ),
+			$content
+		);
+
+		return $html;
+	}
+
+	/**
+	 * Update the context by injecting the correct post data
+	 * for each one of the Single Product inner blocks.
+	 *
+	 * @param array    $context Block context.
+	 * @param array    $block Block attributes.
+	 * @param WP_Block $parent_block Block instance.
+	 *
+	 * @return array Updated block context.
+	 */
+	public function update_context( $context, $block, $parent_block ) {
+		if ( 'woocommerce/single-product' == $block['blockName']
+			&& isset( $block['attrs']['productId'] ) ) {
+				$this->product_id = $block['attrs']['productId'];
+
+				$this->single_product_inner_blocks_names = array_reverse(
+					$this->extract_single_product_inner_block_names( $block )
+				);
+		}
+
+		$this->replace_post_for_single_product_inner_block( $block, $context );
+
+		return $context;
+	}
+
+	/**
+	 * Extract the inner block names for the Single Product block. This way it's possible
+	 * to map all the inner blocks for a Single Product block and manipulate the data as needed.
+	 *
+	 * @param array $block The Single Product block or its inner blocks.
+	 * @param array $result Array of inner block names.
+	 *
+	 * @return array Array containing all the inner block names of a Single Product block.
+	 */
+	protected function extract_single_product_inner_block_names( $block, &$result = [] ) {
+		if ( isset( $block['blockName'] ) ) {
+			$result[] = $block['blockName'];
+		}
+
+		if ( isset( $block['innerBlocks'] ) ) {
+			foreach ( $block['innerBlocks'] as $inner_block ) {
+				$this->extract_single_product_inner_block_names( $inner_block, $result );
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Replace the global post for the Single Product inner blocks and reset it after.
+	 *
+	 * This is needed because some of the inner blocks may use the global post
+	 * instead of fetching the product through the `productId` attribute, so even if the
+	 * `productId` is passed to the inner block, it will still use the global post.
+	 *
+	 * @param array $block Block attributes.
+	 * @param array $context Block context.
+	 */
+	protected function replace_post_for_single_product_inner_block( $block, &$context ) {
+		if ( $this->single_product_inner_blocks_names ) {
+			$block_name = array_pop( $this->single_product_inner_blocks_names );
+
+			if ( $block_name === $block['blockName'] ) {
+				global $post;
+				$post = get_post( $this->product_id );
```

Comment:
> Just fixed them, there is just one related to the global $post that it's basically the one way we found to make the correct product to display for this block, so I cannot edit it

In that case, I think we should add a `phpcs:ignore` directive so the automated job keeps passing.